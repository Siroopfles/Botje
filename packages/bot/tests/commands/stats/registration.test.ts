import { describe, it, expect } from '@jest/globals';
import { 
  SlashCommandBuilder, 
  SlashCommandSubcommandGroupBuilder,
  SlashCommandSubcommandBuilder,
  SlashCommandStringOption,
  SlashCommandUserOption,
  ApplicationCommandOptionType
} from 'discord.js';
import { command } from '../../../src/commands/stats/index.js';

describe('Stats Command Registration', () => {
  describe('command definition', () => {
    const builder = command.data as SlashCommandBuilder;

    it('should match snapshot', () => {
      expect(builder.toJSON()).toMatchSnapshot();
    });

    it('should have correct base properties', () => {
      expect(builder.name).toBe('stats');
      expect(builder.description).toBeDefined();
      expect(builder.description.length).toBeGreaterThan(0);
    });

    it('should require guild context', () => {
      expect(builder.dm_permission).toBeFalsy();
      expect(builder.default_member_permissions).toBeDefined();
    });

    describe('subcommands', () => {
      it('should have permission management group', () => {
        const permissionGroup = builder.options.find(
          opt => opt.toJSON().name === 'permissions'
        ) as SlashCommandSubcommandGroupBuilder;
        
        expect(permissionGroup).toBeDefined();
        const subcommands = permissionGroup.options as SlashCommandSubcommandBuilder[];
        expect(subcommands).toHaveLength(2);
        
        const subcommandNames = subcommands.map(opt => opt.toJSON().name);
        expect(subcommandNames).toContain('view');
        expect(subcommandNames).toContain('reset');
      });

      it('should have task statistics options', () => {
        const taskCommand = builder.options.find(
          opt => opt.toJSON().name === 'tasks'
        ) as SlashCommandSubcommandBuilder;
        
        expect(taskCommand).toBeDefined();
        
        const periodOption = taskCommand.options.find(
          opt => opt.toJSON().name === 'period'
        ) as SlashCommandStringOption;

        expect(periodOption?.required).toBeFalsy();
        expect(periodOption?.choices).toHaveLength(4);
      });

      it('should have user statistics options', () => {
        const userCommand = builder.options.find(
          opt => opt.toJSON().name === 'users'
        ) as SlashCommandSubcommandBuilder;
        
        expect(userCommand).toBeDefined();
        
        const targetOption = userCommand.options.find(
          opt => opt.toJSON().name === 'target'
        ) as SlashCommandUserOption;

        expect(targetOption?.required).toBeFalsy();
      });
    });

    describe('option validation', () => {
      it('should have valid period choices', () => {
        const taskCommand = builder.options.find(
          opt => opt.toJSON().name === 'tasks'
        ) as SlashCommandSubcommandBuilder;

        const periodOption = taskCommand.options.find(
          opt => opt.toJSON().name === 'period'
        ) as SlashCommandStringOption;

        const choices = periodOption.choices?.map(c => ({
          name: c.name,
          value: c.value
        }));

        expect(choices).toEqual([
          { name: 'Today', value: 'day' },
          { name: 'This Week', value: 'week' },
          { name: 'This Month', value: 'month' },
          { name: 'All Time', value: 'all' }
        ]);
      });

      it('should have correct option types', () => {
        const userCommand = builder.options.find(
          opt => opt.toJSON().name === 'users'
        ) as SlashCommandSubcommandBuilder;

        const targetOption = userCommand.options.find(
          opt => opt.toJSON().name === 'target'
        ) as SlashCommandUserOption;

        expect(targetOption?.type).toBe(ApplicationCommandOptionType.User);
      });
    });
  });

  describe('handler registration', () => {
    it('should have execute method', () => {
      expect(command.execute).toBeDefined();
      expect(typeof command.execute).toBe('function');
    });
  });
});