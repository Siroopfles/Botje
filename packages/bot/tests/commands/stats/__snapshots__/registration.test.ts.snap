// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats Command Registration command definition should match snapshot 1`] = `
{
  "default_member_permissions": "8",
  "description": "View various statistics",
  "dm_permission": false,
  "name": "stats",
  "options": [
    {
      "description": "Permission related statistics",
      "name": "permissions",
      "options": [
        {
          "description": "View permission statistics",
          "name": "view",
          "type": 1
        },
        {
          "description": "Reset permission metrics",
          "name": "reset",
          "type": 1
        }
      ],
      "type": 2
    },
    {
      "description": "View task statistics",
      "name": "tasks",
      "options": [
        {
          "choices": [
            {
              "name": "Today",
              "value": "day"
            },
            {
              "name": "This Week",
              "value": "week"
            },
            {
              "name": "This Month",
              "value": "month"
            },
            {
              "name": "All Time",
              "value": "all"
            }
          ],
          "description": "Time period to view stats for",
          "name": "period",
          "required": false,
          "type": 3
        }
      ],
      "type": 1
    },
    {
      "description": "View user activity statistics",
      "name": "users",
      "options": [
        {
          "description": "User to view stats for (defaults to yourself)",
          "name": "target",
          "required": false,
          "type": 6
        }
      ],
      "type": 1
    }
  ]
}